[WIP] Addition of Dora

Prakyathkantharaju opened the pull request:
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
I have added DORA for the loralinear module. Request was proposed in the issue: https://github.com/pytorch/torchtune/issues/893

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help.)

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add unit tests for any new functionality
- [x] update docstrings for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
	- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

ebsmothers commented on torchtune/modules/peft/lora.py:
```diff
@@ -128,6 +134,17 @@ def forward(self, x: Tensor) -> Tensor:
+                + (self.alpha / self.rank)
+                * (self.lora_a.weight.T @ self.lora_b.weight.T).T,
+                dim=1,
+            ).to(self.weight.dtype)
```
What about the detach used in the PEFT implementation referenced in your comment?

ebsmothers commented on torchtune/modules/peft/lora.py:
```diff
@@ -128,6 +134,17 @@ def forward(self, x: Tensor) -> Tensor:
+                dim=1,
+            ).to(self.weight.dtype)
+            mag_norm_scale = (self.m / weight_norm - 1).view(1, -1)
+            return mag_norm_scale * out + mag_norm_scale * lora_out
```
Sorry maybe I am still missing the point but I don't think this actually matches the version that was ultimately added to PEFT. [Ref](https://github.com/huggingface/peft/blob/02ae6bcb373d9d9d3bec9ba920d63316418ff64a/src/peft/tuners/lora/layer.py#L250-L252)

ebsmothers commented on torchtune/modules/peft/lora.py:
```diff
@@ -128,6 +134,17 @@ def forward(self, x: Tensor) -> Tensor:
+            weight_norm = torch.linalg.norm(
+                self.weight
+                + (self.alpha / self.rank)
+                * (self.lora_a.weight.T @ self.lora_b.weight.T).T,
```
Do we need all the transposes here? Isn't this just `self.lora_b.weight @ self.lora_a.weight`?

ebsmothers commented on torchtune/modules/peft/lora.py:
```diff
@@ -67,6 +72,7 @@ def __init__(
         self.dropout = nn.Dropout(p=dropout)
         self.lora_a = nn.Linear(in_features=in_dim, out_features=rank, bias=False)
         self.lora_b = nn.Linear(in_features=rank, out_features=out_dim, bias=False)
+        self.m = nn.Parameter(torch.ones(1, out_dim)) if self.use_dora else None
```
This may necessitate some extra logic for checkpoint loading

ebsmothers reviewed (commented):
Thanks for opening the PR! I am still not sure about the correctness of the implementation though. Can you run forward on the same input tensor and confirm you get the same results with a known correct implementation (e.g. the one from PEFT referenced in your L138 of lora.py).

There are some other forward-looking considerations as well: specifically how we expose DoRA in our higher-level model builders (could potentially be similar to what we do for QLoRA), how we will merge weights when DoRA is applied, determining to what extent we want to support enabling and disabling DoRA adapters (this functionality is used in e.g. our DPO recipe). But for now the main thing is to make sure the linear component itself is correct and well-tested.

rohan-varma commented on tests/torchtune/modules/peft/test_lora.py:
```diff
@@ -97,6 +110,12 @@ def test_forward(self, inputs, lora_linear, out_dim) -> None:
         torch.testing.assert_close(actual.mean(), expected, atol=1e-4, rtol=1e-6)
 
+    def test_dora_forward(self, inputs, dora_linear, out_dim) -> None:
+        expected = torch.tensor(EXPECTED_VAL)
```
Does this mean the expected val of DoRA is the same as LoRA? Why is this, and intuitively, I'm not sure if I understand if the result is the exact same, how DoRA results in different training than LoRA? Pretty sure I'm missing something basic here but would be good to clarify.

Prakyathkantharaju replied:
Here is the explanation:
$$DORA = W*x + (\frac{m}{weight norm} -1)*W *x  + \frac{m}{weight norm} * {lora}_b({lora}_a(x)) + {scaling}$$

The m vector is initialized as self.m == weight norm. So the ratio $\frac{m}{weight norm}$ is 1 for the first iteration.
so LORA == DORA for the first pass.

Better explanation from the author: https://github.com/huggingface/peft/pull/1474#issuecomment-1964581081

rohan-varma commented on torchtune/modules/peft/lora.py:
```diff
@@ -32,6 +33,8 @@ class LoRALinear(nn.Module, AdapterModule):
         rank (int): rank of the low-rank approximation
         alpha (float): scaling factor for the low-rank approximation
         dropout (float): dropout probability. Default: 0.0
+        use_dora (bool): whether to use DORA (weight-Decomposed Low-Rank Adaptation).
```
nit: link to paper.

Prakyathkantharaju replied:
Updated in the new commit.

rohan-varma commented on torchtune/modules/peft/lora.py:
```diff
@@ -32,6 +33,8 @@ class LoRALinear(nn.Module, AdapterModule):
         alpha (float): scaling factor for the low-rank approximation
         dropout (float): dropout probability. Default: 0.0
+        use_dora (bool): whether to use DORA (weight-Decomposed Low-Rank Adaptation).
+            Default: False
```
how do we want to expose DoRA? I see following the pattern of how we enabled QLoRA and just passing in a use_dora flag. Curious about the tradeoffs compared to a `DoraLinear` layer though. @ebsmothers any thoughts?

rohan-varma reviewed (commented):
In addition to @ebsmothers comments, let's have comprehensive testing to ensure we have implementation parity compared to a well-known implementation, such as the one offered by HF PEFT. This will ensure we have the appropriate correctness guarantees in place.

Thanks so much for working on this!

Prakyathkantharaju commented on torchtune/modules/peft/lora.py:
```diff
@@ -111,6 +118,19 @@ def adapter_params(self) -> List[str]:
         adapter_params = ["lora_a.weight", "lora_b.weight"]
         return adapter_params
 
+    def dora_init(self) -> None:
```
Hello @rohan-varma and @ebsmothers, I am wondering if there is a more efficient approach to the problem I am facing. I have followed the PEFT (Parameter Efficient Fine-Tuning) method but I am encountering an issue with the initialization of the self.m (magnitude vector) in torch tune because I require the weights. Currently, I am using the flag (self.dora_initialized) to address this issue, but I believe this may not be the best solution. I would appreciate any suggestions you may have.

Prakyathkantharaju commented:
> Thanks for opening the PR! I am still not sure about the correctness of the implementation though. Can you run forward on the same input tensor and confirm you get the same results with a known correct implementation (e.g. the one from PEFT referenced in your L138 of lora.py).
> 
> There are some other forward-looking considerations as well: specifically how we expose DoRA in our higher-level model builders (could potentially be similar to what we do for QLoRA), how we will merge weights when DoRA is applied, determining to what extent we want to support enabling and disabling DoRA adapters (this functionality is used in e.g. our DPO recipe). But for now the main thing is to make sure the linear component itself is correct and well-tested.

Hello @ebsmothers , There were some bugs in my initial commit, I have fixed those bugs ( detach missing, too many transposes and -1 in the` mag_norm_scale` calculation). I also added a dora_init function, which will initialize the `self.m` vector, similar to PEFT, However, I think this method might not be the best approach. If you have a better idea for initialization, please let me know.

ebsmothers commented:
Hi @Prakyathkantharaju sorry for the delay in responding here. I think beyond just exposing the DoRA logic in LoRALinear we probably want to think about the overall design, interaction with other parts of the library, and thorough testing (basically some of the points in my comment [here](https://github.com/pytorch/torchtune/pull/936#pullrequestreview-2039844350) and @rohan-varma's comment [here](https://github.com/pytorch/torchtune/pull/936#pullrequestreview-2040212285)). Since it's a fair amount of effort to do all of this, I am gonna tag in @calvinpelletier to help out here. Let me know if this works for you, we'd love to have your collaboration on design and code reviews.

Prakyathkantharaju commented:
Hello and thank you for your response. I apologize for not updating this issue for a long time. I am currently working on comparing the performance of the Dora implementation with the PEFT. This was a missing aspect from the request raised by @ebsmothers and @rohan-varma. For this test, I plan to use the LLAMA-3 single GPU Lora. Over the weekend, I will provide more information about the code used for comparison and the results. Thank you for your patience. As you suggested, I am okay and willing to collaborate with @calvinpelletier and anyone else who is willing to. Let me know if you have any other request.

Prakyathkantharaju commented on recipes/configs/llama3/8B_Dora_single_device.yml:
```diff
@@ -0,0 +1,85 @@
+  apply_lora_to_mlp: True
+  apply_lora_to_output: False
+  lora_rank: 8
+  lora_alpha: 16
```
I have updated the configs and recipe now there is no Dora recipe, and everything is done using the Lora recipe, where I check if use_dora is defined and perform a specific class.

Prakyathkantharaju commented on recipes/dora_finetune_single_device.py:
```diff
@@ -0,0 +1,557 @@
+log = utils.get_logger("DEBUG")
+
+
+class LoRAFinetuneRecipeSingleDevice(FTRecipeInterface):
```
Update. Now, everything is done through Lora recipe now.

Prakyathkantharaju commented on torchtune/modules/peft/lora.py:
```diff
@@ -143,3 +183,10 @@ def _lora_b_init_params(x: nn.Linear) -> None:
+    """
+    Initialize DORA m to ones.
+    """
+    nn.init.zeros_(x)
```
Updated with ones. These weights are updated later in the init_dora code. I wanted to keep the initialization of the new weights very similar to Lora, so I followed this format.

Prakyathkantharaju commented on torchtune/modules/peft/lora.py:
```diff
@@ -109,8 +116,34 @@ def adapter_params(self) -> List[str]:
+    def _dora_weight_norm(self) -> Tensor:
+        if self._quantize_base:
+            # Convert NF4Tensor to regular Tensor for computation TODO(prakyath): Fix this.
+            weight = to_regular_tensor(self.weight)
```
Updated, now the weight norm is calculated by dequantizing the weights. Please let me know if you have a better/faster way to do this ?

Prakyathkantharaju commented:
Hello everyone,

I apologize for the delayed response, and I appreciate your review of my changes.

I have addressed the comments made by @ebsmothers and updated the structure of how Dora is initialized. Here are the details on how Dora is initialized (I have kept the initialization as similar to QLora as possible):

1. I added a partial Dora class, similar to QLora, where Dora is initialized by the `use_dora` option. You can find the link to the change [here](https://github.com/pytorch/torchtune/blob/9110f0e1e3f69c61a71f66fe85f617de35dd8300/torchtune/models/llama3/_model_builders.py#L186).
2. I updated the Lora recipe with Dora initialization. You can find the link to the changes [here](https://github.com/pytorch/torchtune/blob/733aff4f8c49dc59282c7d6f57cbe6c41e94ff3b/recipes/lora_finetune_single_device.py#L279).
3. I updated the llama-3 lora initialization with the Dora-specific option. The links to the changed files are [here](https://github.com/pytorch/torchtune/blob/0ad3b8492bceab5f58bc8446b1840b7fca4b5da5/torchtune/models/llama3/_component_builders.py#L154).

Additionally, I would like feedback from @rohan-varma  @kostmo and @ebsmothers. I also added a clamp function to avoid making the denominator 0 (link to the code [here](https://github.com/pytorch/torchtune/blob/0ad3b8492bceab5f58bc8446b1840b7fca4b5da5/torchtune/modules/peft/lora.py#L149)). This is not done in the peft or other Dora implementations. If you feel this is not necessary, then I can remove it.

Please let me know if you need any changes. I am willing to work on them. Moreover, if you have any additional feedback, I am happy to incorporate it as well.

calvinpelletier reviewed (commented):
Hi @Prakyathkantharaju, thanks for contributing! I left some comments.

One thing that you are currently missing is updating the merging logic at `torchtune/module/peft/peft_utils.py::get_merged_lora_ckpt`.

I'll work on setting up a direct comparison to the huggingface implementation to verify that the loss graph, memory usage, and training speed are similar.

Prakyathkantharaju commented on torchtune/_recipe_registry.py:
```diff
@@ -106,6 +106,10 @@ class Recipe:
                 file_path="gemma/2B_qlora_single_device.yaml",
             ),
+            Config(
+                name="llama3/*B_dora_single_device",
```
Updated

Prakyathkantharaju commented on torchtune/models/llama3/_model_builders.py:
```diff
@@ -180,3 +183,12 @@ def lora_llama3_70b(
+
+dora_llama3_8b.__doc__ = """
+Builder for creating a Llama3 model with DORA enabled. Base model weights in linear layers
+that DORA is applied to are quantized per the Dora paper: https://arxiv.org/abs/2402.09353.
```
I have fixed these; I will check again before the merge to ensure that all the instances of dora/Dora are changed to DoRA.

Prakyathkantharaju commented on recipes/configs/llama3/8B_dora_single_device.yaml:
```diff
@@ -0,0 +1,86 @@
+# You can add specific overrides through the command line. For example
+# to override the checkpointer directory while launching training
+# you can run:
+#   tune run lora_finetune_single_device --config llama3/8b_qlora_single_device checkpointer.checkpoint_dir=<YOUR_CHECKPOINT_DIR>
```
Updated in the latest push.

Prakyathkantharaju commented on recipes/configs/llama3/8B_dora_single_device.yaml:
```diff
@@ -0,0 +1,86 @@
+# Tokenizer
+tokenizer:
+  _component_: torchtune.models.llama3.llama3_tokenizer
+  path: /teamspace/studios/this_studio/models/Meta-Llama-3-8b-Instruct/original/tokenizer.model
```
I have updated the logging to disk logger; I will update this path to /tmp/ before the merge.

Prakyathkantharaju commented on torchtune/modules/peft/lora.py:
```diff
@@ -109,8 +118,38 @@ def adapter_params(self) -> List[str]:
+        self.lora_m.data = weight_norm.data  # Update the data of 'm' directly
+
+    @property
+    def _dora_weight_norm(self) -> Tensor:
```
I update this to be a function.

Prakyathkantharaju commented on torchtune/modules/peft/peft_utils.py:
```diff
@@ -63,6 +63,13 @@ def get_adapter_params(model: nn.Module) -> Dict[str, nn.Parameter]:
+def activate_dora_params(model: nn.Module) -> nn.Module:
+    for k, v in model.named_modules():
+        if hasattr(v, "adapter_params") and callable(v.adapter_params):
+            current_adapter_params = v.adapter_params()
```
Removed in the latest commit

Prakyathkantharaju commented on torchtune/modules/peft/lora.py:
```diff
@@ -109,8 +118,38 @@ def adapter_params(self) -> List[str]:
+        norm = torch.linalg.norm(result, dim=1)
+
+        # Clamp the norm to avoid division by zero
+        # TODO(Prakyath): Check with torchtune team whether this should be a parameter ?
```
The comment was referring to the clamp line [here](https://github.com/Prakyathkantharaju/torchtune/blob/7b4b8a4dbae45f5b1b436fe902c6ec5006e3e28c/torchtune/modules/peft/lora.py#L147). I have made the minimum of the clamp 1e-6. I was thinking this could be a parameter configured in the yaml file. FYI: I added this clamp to ensure that the denominator of this line is not 0. This is not present in the peft.

Prakyathkantharaju commented:
@calvinpelletier Thank you for reviewing. You are right, I have to implement `torchtune/module/peft/peft_utils.py::get_merged_lora_ckpt` However, for dora merging, since it involves a weight normalization ([link](https://github.com/Prakyathkantharaju/torchtune/blob/7b4b8a4dbae45f5b1b436fe902c6ec5006e3e28c/torchtune/modules/peft/lora.py#L138C23-L138C29)) I need the linear layer weight in addition to the lora_a and lora_b weights. For this, I need to change the recipe, and I need to write a new merge function. I want to clarify this before I start editing the recipe since, in the previous conversation, there was hesitation in updating the recipe. If you have a different approach in mind, I am open to that as well.
Regards,
PK

RdoubleA commented:
covered by #1115
